### PR TITLE
Merkle Proof Support for BeaconState Finalized Root and Sync Committees

### PR DESCRIPTION
**Pull request by rauljordan**
An important part for light-client support is being able to retrieve proofs from the beacon state and verify them. A required item we want to prove is the finalized block root contained in the finalized checkpoint of a beacon state. This PR adds support for this, which will be immediately impactful to light client prototypes. cc @jinfwhuang.

Unfortunately, this code is as deduplicated as I could make it given our current design pattern for v1, v2, and v3 beacon states.

### CODE REVIEW

File: beacon-chain/state/v1/unsupported_getters.go [hunk 1]
```diff
@@ -34,3 +34,13 @@ func (*BeaconState) NextSyncCommittee() (*ethpb.SyncCommittee, error) {
 func (*BeaconState) LatestExecutionPayloadHeader() (*ethpb.ExecutionPayloadHeader, error) {
 	return nil, errors.New("LatestExecutionPayloadHeader is not supported for phase 0 beacon state")
 }
+
+// ProveFinalizedRoot --
+func (b *BeaconState) ProveFinalizedRoot() ([][]byte, error) {
```
Comment (rauljordan) [inline beacon-chain/state/v1/unsupported_getters.go:39]: This should work for v1, but not the sync committee proofs, so just laying out this pattern once that is added

File: container/trie/sparse_merkle.go [hunk 2]
```diff
@@ -197,6 +197,20 @@ func VerifyMerkleBranch(root, item []byte, merkleIndex int, proof [][]byte, dept
 	return bytes.Equal(root, node[:])
 }
 
+// VerifyMerkleProof given a trie root, a leaf, the generalized merkle index
+// of the leaf in the trie, and the proof itself.
+func VerifyMerkleProof(root, item []byte, merkleIndex int, proof [][]byte) bool {
```
Comment (rauljordan) [inline container/trie/sparse_merkle.go:202]: We don't need a depth to verify merkle proofs for the state, but need a depth for verifying deposit merkle proofs, so I separate the methods

File: beacon-chain/state/v2/proofs.go [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+// FinalizedRootGeneralizedIndex for the Altair beacon state.
+func FinalizedRootGeneralizedIndex() int {
+	return finalizedRootIndex
+}
+
+// ProveFinalizedRoot crafts a Merkle proof for the finalized root
+// contained within the finalized checkpoint of a beacon state.
+func (b *BeaconState) ProveFinalizedRoot() ([][]byte, error) {
+	cpt := b.state.FinalizedCheckpoint
```
Comment (terencechain) [inline beacon-chain/state/v2/proofs.go:21]: Don't we need locks here?

File: beacon-chain/state/v2/proofs.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	epochBuf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(epochBuf, uint64(cpt.Epoch))
+	epochRoot := bytesutil.ToBytes32(epochBuf)
```
Comment (terencechain) [inline beacon-chain/state/v2/proofs.go:24]: The topmost root is derived from epoch number? Interesting

File: beacon-chain/state/phase0.go [hunk 5]
```diff
@@ -18,6 +18,12 @@ type BeaconState interface {
 	Copy() BeaconState
 	HashTreeRoot(ctx context.Context) ([32]byte, error)
 	FutureForkStub
+	StateProver
+}
+
+// StateProver defines the ability to create Merkle proofs for beacon state fields.
+type StateProver interface {
+	ProveFinalizedRoot() ([][]byte, error)
```
Comment (terencechain) [inline beacon-chain/state/phase0.go:26]: Optional:

```suggestion
	FinalizedRootProofs() ([][]byte, error)
```

File: beacon-chain/state/v2/proofs.go (hunk 4 above)
Comment (rauljordan) [reply beacon-chain/state/v2/proofs.go:24]: no, but the epoch is the neighbor of the finalized root field, so the proof consists of walking up the following branch

```
    state_root
                \
             0x213912
                    \
               htr(finalized_checkpoint)
                        /              \
                  epoch           root
```

File: beacon-chain/state/phase0.go (hunk 5 above)
Comment (rauljordan) [reply beacon-chain/state/phase0.go:26]: It's a single proof, maybe I can just call it FinalizedRootProof, but Prove as a prefix is a bit easier to read and immediately tells you what it does

Comment (rauljordan) [reply beacon-chain/state/phase0.go:26]: All future proof methods can have the Prove prefix, so ProveNextSyncCommitte, etc.

Comment (terencechain) [reply beacon-chain/state/phase0.go:26]: `ProveFinalizedRoot` makes me think of something like `ProveFinalizedRoot([][]byte) (bool, error)` given the prefix is a verb. I think this is intended to be a getter. No hard preference for either. You choose

**Comment by rauljordan** [comment]
Finally ready @terencechain. Supports all the fields we need for proofs in the light client!

**Review by terencechain — APPROVED** [review]

**Pull request merged**